datastruct/mpool.h

```diff
@@ -11,69 +13,114 @@
  * of allowing the memory to be reused by some other code.
  */
 
+struct mpool {
+	size_t stacklen;
+	size_t allocsize;
+	void ** allocs;
+	uint64_t nallocs;
+	uint64_t nempties;
+	int state;
+	void ** allocs_static;
+	void (*atexitfunc)(void);
+};
+
+static inline void
+mpool_atexit(struct mpool * M)
+{
+
+	while (M->stacklen)
+		free(M->allocs[--M->stacklen]);
+	if (M->allocs != M->allocs_static)
+		free(M->allocs);
+}
+
+static inline void *
+mpool_malloc(struct mpool * M, size_t len)
+{
+
+	M->nallocs++;
+	if (M->stacklen)
```

Comment:
(follow-up to my comment about `M->stacklen`)

... because if `M->stacklen==0`, then I don't think we ever enter this `if`.

Reply:
If `stacklen == 0` then we don't have any allocations in the pool.  So we allocate a new one.  But `stacklen` can increase in mpool_free.